### PR TITLE
Implement `CircuitDisbandRequest` in admin service

### PR DESCRIPTION
**Pull request by shannynalayna**
Adds a few commits necessary to implement the `CircuitDisbandRequest` in the admin service. 

This feature will be fully testable once the CLI commands are implemented (wip), but otherwise run `cargo test --features experimental disband` to specify the tests added for this implementation.

### CODE REVIEW

File: libsplinter/src/admin/service/mod.rs [hunk 1]
```diff
@@ -1089,6 +1090,214 @@ mod tests {
         mesh.shutdown_signaler().shutdown();
     }
 
+    /// Test that a circuit disband sends the appropriate messages.
```
Comment (agunde406) [inline libsplinter/src/admin/service/mod.rs:1093]: Can you expand this description so it explains the test more.

File: libsplinter/src/admin/service/shared.rs [hunk 2]
```diff
@@ -390,89 +391,143 @@ impl AdminServiceShared {
...
+                                        "Unable to remove circuit from routing table: {}",
+                                        circuit_id
+                                    ))
+                                })?;
+                            // Removing the circuit's peer refs
+                            for member in circuit_proposal.get_circuit_proposal().get_members() {
+                                self.remove_peer_ref(member.get_node_id());
+                            }
+                            // Removing the corresponding `active` circuit from the admin store
+                            self.admin_store.remove_circuit(circuit_id).map_err(|_| {
```
Comment (agunde406) [inline libsplinter/src/admin/service/shared.rs:498]: We should use update_circuit here instead of remove/upgrade to keep the operation atomic. This will require you get the circuit from the proposal here instead of relying on the admin service store to do the conversion.

File: libsplinter/src/admin/service/shared.rs [hunk 3]
```diff
@@ -506,6 +561,18 @@ impl AdminServiceShared {
...
+                            CircuitManagementPayload_Action::CIRCUIT_DISBAND_REQUEST => {
+                                // notify registered application authorization handlers of the
+                                // committed disband circuit proposal
+                                let event = messages::AdminServiceEvent::ProposalSubmitted(
+                                    messages::CircuitProposal::from_proto(circuit_proposal.clone())
+                                        .map_err(AdminSharedError::InvalidMessageFormat)?,
+                                );
+                                self.send_event(&mgmt_type, event);
+
+                                info!("committed changes to disband circuit {}", circuit_id);
```
Comment (agunde406) [inline libsplinter/src/admin/service/shared.rs:573]: This should say `info!("committed changes for new circuit proposal to disband circuit {}", circuit_id);` and the create log above should say to `proposal to create circuit {}`

File: libsplinter/src/admin/service/shared.rs [hunk 4]
```diff
@@ -913,6 +1062,69 @@ impl AdminServiceShared {
...
+        message_sender: String,
+    ) -> Result<(), ServiceError> {
+        let mut missing_protocol_ids = vec![];
+        let mut pending_peers = vec![];
+        let mut pending_members = vec![];
+        let mut added_peers: Vec<String> = vec![];
+        for node in members {
+            if self.node_id() != node.get_node_id() {
+                debug!("Referencing node {:?}", node);
+                let peer_ref = self
```
Comment (agunde406) [inline libsplinter/src/admin/service/shared.rs:1078]: This should not be adding peer_refs, required peer_refs will have been added when the circuit was created. This is adding an extra set of peer_refs that will not be removed when the disband request is approved. Only create_payloads should add new peer_refs

File: libsplinter/src/admin/service/shared.rs [hunk 5]
```diff
@@ -1500,12 +1795,6 @@ impl AdminServiceShared {
                     return Err(AdminSharedError::ValidationFailed(
                         "Proposed circuit cannot have a display name on protocol 1".to_string(),
                     ));
-                } else if circuit.get_circuit_status()
```
Comment (agunde406) [inline libsplinter/src/admin/service/shared.rs:1503]: why was this deleted?

File: libsplinter/src/admin/service/shared.rs [hunk 6]
```diff
@@ -1824,6 +2113,81 @@ impl AdminServiceShared {
...
+                AdminSharedError::ValidationFailed(format!(
+                    "{} is not permitted to vote for node {}",
+                    to_hex(signer_public_key),
+                    requester_node_id
+                ))
+            })?;
+
+        if self.has_proposal(circuit.get_circuit_id())? {
+            return Err(AdminSharedError::ValidationFailed(format!(
+                "Ignoring duplicate disband proposal of circuit {}",
```
Comment (agunde406) [inline libsplinter/src/admin/service/shared.rs:2153]: This should be updated to say "Ignoring duplicate proposal for circuit" (and in the create validation) since it might be a different proposal type

**Review by agunde406 — CHANGES_REQUESTED** [review]

File: libsplinter/src/admin/service/mod.rs (hunk 1 above)
Comment (agunde406) [reply libsplinter/src/admin/service/mod.rs:1093]: same for other tests added

Comment (shannynalayna) [reply libsplinter/src/admin/service/mod.rs:1093]: Yep!

File: libsplinter/src/admin/service/shared.rs (hunk 5 above)
Comment (shannynalayna) [reply libsplinter/src/admin/service/shared.rs:1503]: This is checking the protocol version, rather than the `circuit_version` (This was an error I was initially running into while testing the disband request). I added a check to verify the circuit version in the `validate_disband_request` rather than the admin service protocol, as it is doing here.

File: libsplinter/src/admin/service/shared.rs [hunk 7]
```diff
@@ -487,7 +552,11 @@ impl AdminServiceShared {
                                 );
                                 self.send_event(&mgmt_type, event);
 
-                                info!("committed changes for new circuit proposal {}", circuit_id);
+                                info!(
+                                    "committed changes for new ciircuit proposal to create circuit \
```
Comment (agunde406) [inline libsplinter/src/admin/service/shared.rs:556]: typo ciircuit -> circuit

File: libsplinter/src/admin/service/shared.rs [hunk 8]
```diff
@@ -705,6 +789,66 @@ impl AdminServiceShared {
...
+                let mut protocol = ADMIN_SERVICE_PROTOCOL_VERSION;
+                for member in circuit_proposal.get_circuit_proposal().get_members() {
+                    verifiers.push(admin_service_id(member.get_node_id()));
+                    // Figure out what protocol version should be used for this proposal
+                    if let Some(protocol_version) = self
+                        .service_protocols
+                        .get(&admin_service_id(member.get_node_id()))
+                    {
+                        if protocol_version < &protocol {
+                            protocol = *protocol_version
```
Comment (agunde406) [inline libsplinter/src/admin/service/shared.rs:829]: This protocol isnt used. If the protocol version is 1, it should reject this proposal right away since Disband is not supported.

**Review by agunde406 — CHANGES_REQUESTED** [review]
This should be behind an experimental feature.

File: libsplinter/src/admin/service/shared.rs (hunk 7 above)
Comment (shannynalayna) [reply libsplinter/src/admin/service/shared.rs]: Fixed

File: libsplinter/src/admin/service/shared.rs (hunk 8 above)
Comment (shannynalayna) [reply libsplinter/src/admin/service/shared.rs:844]: Fixed to send this protocol to the `validate_disband_circuit` method, and added a check to this method for the protocol (returning an error if it's 1) and added a test `test_validate_disband_circuit_invalid_protocol` to validate this.

**Review by agunde406 — APPROVED** [review]

File: libsplinter/src/admin/service/messages/mod.rs [hunk 9]
```diff
@@ -210,6 +210,75 @@ impl CreateCircuit {
 
         Ok(create_request)
     }
+
+    pub fn into_circuit_proto(self) -> Result<admin::Circuit, MarshallingError> {
```
Comment (peterschwarz) [inline libsplinter/src/admin/service/messages/mod.rs:214]: This seems a bit odd to include here, in a struct that transforms to several different types. Perhaps in a future PR, this should be moved to a `TryInto` implementation, or some other strategy.

File: libsplinter/src/admin/service/error.rs [hunk 10]
```diff
@@ -161,6 +161,10 @@ pub enum AdminSharedError {
         context: String,
         source: Option<InitializeServiceError>,
     },
+    ServiceShutdownFailed {
```
Comment (peterschwarz) [inline libsplinter/src/admin/service/error.rs:167]: This enum should be reduced to have the types defined in the `crate::error` module.  Another note to make for a future PR

File: libsplinter/src/admin/store/circuit.rs [hunk 11]
```diff
@@ -92,6 +95,56 @@ impl Circuit {
     }
 }
 
+impl TryFrom<&admin::Circuit> for Circuit {
+    type Error = InvalidStateError;
+
+    fn try_from(proto: &admin::Circuit) -> Result<Self, Self::Error> {
```
Comment (peterschwarz) [inline libsplinter/src/admin/store/circuit.rs:101]: This should be in a separate commit.

Comment (shannynalayna) [reply libsplinter/src/admin/store/circuit.rs:101]: Addressed

File: libsplinter/src/admin/service/messages/mod.rs (hunk 9 above)
Comment (shannynalayna) [reply libsplinter/src/admin/service/messages/mod.rs:214]: I added these comments (besides the separate commit one; I addressed in the PR) to the `circuit-disband` feature doc so these will get revisited in the stabilization process

**Review by agunde406 — APPROVED** [review]

**Review by peterschwarz — APPROVED** [review]
